### PR TITLE
envoy: Pass 'non-redirect' http traffic through.

### PR DESCRIPTION
**Pull request by jrajahalme**
Cilium only creates a host proxy redirect when the relevant policy has
http rules. A sidecar proxy gets also traffic that would not have been
redirected with a host proxy. To emulate the host proxy behavior we
pass (allow) traffic through a sidecar if it is passed through by the
bpf datapath and would not have been redirected to a host proxy.

There are two cases in cilium network policy enforcement where we can
detect that the given packet would not have been forwarded to a host
proxy:

1. There is no port policy that covers this connection. Cilium always
configures a port policy for a redirect, and hence if we find that
there is no applicable port policy, the connection must have been
passed by the bpf datapath to a sidecar and the sidecar proxy should
pass it through, i.e., consider the decision made by the bpf datapath
final.

2. There is a port policy, but it does not have any http rules. Again,
in this case this request would not have been redirected to a host
proxy, and we must consider bpf datapath policy decision as final by
passing the request through the sidecar proxy.

There changes are marked with TODOs as they will need to be
reconsidered when a non-bpf datapaths are supported.

Suggested-by: Romain Lenglet <romain@covalent.io>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

### CODE REVIEW

**Comment by jrajahalme** [comment]
test-me-please

File: envoy/cilium_network_policy.h [hunk 1]
```diff
@@ -109,16 +109,26 @@ class NetworkPolicyMap : public Singleton::Instance,
...
+	    have_http_rules_ = true;
+	  }
 	  rules_.emplace_back(PortNetworkPolicyRule(it));
 	}
       }
 
       bool Matches(uint64_t remote_id, const Envoy::Http::HeaderMap& headers) const {
+	if (!have_http_rules_) {
+	  // If there are no L7 rules, host proxy will not create a proxy redirect at all,
+	  // whereby the decicion made by the bpf datapath is final. Emulate the same behavior
```
Comment (rlenglet) [inline envoy/cilium_network_policy.h:127]: decicion -> decision

File: envoy/cilium_network_policy.h [hunk 2]
```diff
@@ -109,16 +109,26 @@ class NetworkPolicyMap : public Singleton::Instance,
...
+	  // in the sidecar by allowing such traffic.
+	  // TODO: This will need to be revised when non-bpf datapaths are to be supported.
```
Comment (rlenglet) [inline envoy/cilium_network_policy.h:129]: revised -> removed

File: envoy/cilium_network_policy.h [hunk 3]
```diff
@@ -152,17 +163,28 @@ class NetworkPolicyMap : public Singleton::Instance,
...
+	    return true;
+	  }
+	  found_port_rule = true;
 	}
-	return false;
+
+	// No policy for the port was found. Cilium always creates a policy for redirects it
+	// creates, so the host proxy never gets here. Sidecar gets all the traffic, which we need
+	// to pass through since the bpf datapath already allowed it.
+        // TODO: Change back to false only when non-bpf datapath is supported?
```
Comment (rlenglet) [inline envoy/cilium_network_policy.h:186]: Fix indentation.

Comment (rlenglet) [inline envoy/cilium_network_policy.h:186]:
```
// TODO: Change back to always returning false ("default deny") when non-bpf datapath is supported?
```

**Review by rlenglet — APPROVED** [review]
LGTM! Just a few minor nitpicks.

File: envoy/cilium_network_policy.h [hunk 4]
```diff
@@ -152,17 +163,28 @@ class NetworkPolicyMap : public Singleton::Instance,
       }
 
       bool Matches(uint32_t port, uint64_t remote_id, const Envoy::Http::HeaderMap& headers) const {
+	bool found_port_rule = false;
```
Comment (rlenglet) [inline envoy/cilium_network_policy.h:166]: I don't understand why this is necessary.
I think the case when the port was not found was already handled by the final `return` at the bottom.
That `return` is hit only when the port is not found.
The change was already correct, you only need to change the value of the last `return` to `true`.

**Review by rlenglet — CHANGES_REQUESTED** [review]
Please hold on merging this for now. This may break policy semantics because of another bug we just found in the datapath.

**Review by rlenglet — APPROVED** [review]

**Comment by borkmann** [comment]
Still waiting on the minor nits reported by @rlenglet to be fixed. Thx

**Comment by rlenglet** [comment]
We’re good. Let’s merge this.

**Pull request merged**